arraycontext/container/traversal.py

```diff
@@ -493,6 +501,97 @@ def thaw(ary: ArrayOrContainerT, actx: ArrayContext) -> ArrayOrContainerT:
 # }}}
 
 
+# {{{ flatten / unflatten
+
+def flatten(ary: ArrayOrContainerT, actx: ArrayContext) -> Any:
+    """Convert all arrays in the :class:`~arraycontext.ArrayContainer`
+    into single flat array of a type :attr:`arraycontext.ArrayContext.array_types`.
+
+    The operation requires :attr:`arraycontext.ArrayContext.np` to have
+    ``ravel`` and ``concatenate`` methods implemented. The order in which the
+    individual leaf arrays appear in the final array is dependent on the order
+    given by :func:`~arraycontext.serialize_container`.
+    """
+    @memoize_in(actx, (flatten, "ravel_prg"))
+    def _ravel_prg(shape: Tuple[int, ...]) -> Any:
+        raise NotImplementedError
+
+    def _flatten(subary: ArrayOrContainerT) -> None:
+        try:
+            iterable = serialize_container(subary)
+        except TypeError:
+            try:
+                flat_subary = actx.np.ravel(subary, order="A")
+            except ValueError:
+                flat_subary = actx.call_loopy(
+                        _ravel_prg(subary.shape), ary=subary)
+
+            result.append(flat_subary)
+        else:
+            for _, isubary in iterable:
+                _flatten(isubary)
+
+    result: List[Any] = []
+    _flatten(ary)
+
+    return actx.np.concatenate(result)
+
+
+def unflatten(
+        template: ArrayOrContainerT, ary: Any,
+        actx: ArrayContext) -> ArrayOrContainerT:
+    """Unflatten an array produced by :func:`flatten` back into an
+    :class:`~arraycontext.ArrayContainer`.
+
+    The order and sizes of each slice into *ary* are determined by the
+    array container *template*.
+    """
+    # NOTE: https://github.com/python/mypy/issues/7057
+    offset = 0
+
+    @memoize_in(actx, (unflatten, "reshape_prg"))
+    def _reshape_prg(shape: Tuple[int, ...]) -> Any:
+        raise NotImplementedError
+
+    def _unflatten(template_subary: ArrayOrContainerT) -> ArrayOrContainerT:
+        nonlocal offset
+
+        try:
+            iterable = serialize_container(template_subary)
+        except TypeError:
+            # NOTE: the max is needed to handle device scalars with size == 0
+            offset += max(1, template_subary.size)
+            if offset > ary.size:
+                raise ValueError("'template' and 'ary' sizes do not match")
+
+            flat_subary = ary[offset - template_subary.size:offset]
+            try:
+                subary = actx.np.reshape(flat_subary, template_subary.shape)
+            except ValueError:
+                subary = actx.call_loopy(
+                        _reshape_prg(template_subary.shape), ary=flat_subary)
+
+            return actx.np.astype(subary, template_subary.dtype)
```

Comment:
This reveals another somewhat unanswered question of this interface. Are we serializing into a string of bytes, or into a string of (dtyped) numbers? If the former, this is reasonable. If the latter, there's not really a need to force a dtype on an already dtyped array. (I *think* what we're shooting for is the latter, but I could be convinced. E.g. "unflatten this complex-valued array to look just like this real-valued one" seems like a somewhat reasonable request.

Reply:
The scenario I had in mind here was an array container that had both complex and real arrays in it somewhere, for whatever reason. Then `concatenate` would force that into complex (?), which is not great.

Maybe we should forbid that case for now?

Reply:
I'm OK with restricting the arrays to be dtype-homogeneous for now. That said, I wouldn't find it objectionable if the concatenation coerced things to complex. I kind of think of array containers as "one array" anyhow, just with additional internal structure.

Reply:
Added some extra checks in dd75ade.

My worry was more about the "unflattening" part, since not casting to the dtype from `template` would mean that the operation isn't quite the reverse of "flattening".

Reply:
I want to echo @alexfikl's concern. Generally speaking, I would hope that `unflatten` and `flatten` would be "inverses" of one another if the template is the original container type. If not, this just sounds like an obscure bug waiting to happen.